backends/ref/ceed-ref-operator.c

```diff
@@ -434,6 +534,146 @@ static int CeedCompositeOperatorApply_Ref(CeedOperator op, CeedVector invec,
   return 0;
 }
 
+// Assemble Linear QFunction
+static int CeedOperatorAssembleLinearQFunction_Ref(CeedOperator op,
+    CeedVector assembled, CeedElemRestriction *rstr, CeedRequest *request) {
+  int ierr;
+  CeedOperator_Ref *impl;
+  ierr = CeedOperatorGetData(op, (void *)&impl); CeedChk(ierr);
+  CeedQFunction qf;
+  ierr = CeedOperatorGetQFunction(op, &qf); CeedChk(ierr);
+  CeedInt Q, numelements, numinputfields, numoutputfields, size;
+  ierr = CeedOperatorGetNumQuadraturePoints(op, &Q); CeedChk(ierr);
+  ierr = CeedOperatorGetNumElements(op, &numelements); CeedChk(ierr);
+  ierr= CeedQFunctionGetNumArgs(qf, &numinputfields, &numoutputfields);
+  CeedChk(ierr);
+  CeedOperatorField *opinputfields, *opoutputfields;
+  ierr = CeedOperatorGetFields(op, &opinputfields, &opoutputfields);
+  CeedChk(ierr);
+  CeedQFunctionField *qfinputfields, *qfoutputfields;
+  ierr = CeedQFunctionGetFields(qf, &qfinputfields, &qfoutputfields);
+  CeedChk(ierr);
+  CeedVector vec;
+  CeedInt numactivein = 0, numactiveout = 0;
+  CeedScalar **activein = NULL;
+  CeedScalar *a, *tmp;
+
+  // Setup
+  ierr = CeedOperatorSetup_Ref(op); CeedChk(ierr);
+  ierr = CeedVectorGetArray(assembled, CEED_MEM_HOST, &a); CeedChk(ierr);
+
+  // Input Evecs and Restriction
+  ierr = CeedOperatorSetupInputs_Ref(numinputfields, qfinputfields,
+    opinputfields, NULL, true, impl, request); CeedChk(ierr);
+
+  // Count number of active input fields
+  for (CeedInt i=0; i<numinputfields; i++) {
+    // Get input vector
+    ierr = CeedOperatorFieldGetVector(opinputfields[i], &vec); CeedChk(ierr);
+    // Check if active input
+    if (vec == CEED_VECTOR_ACTIVE) {
+      ierr = CeedQFunctionFieldGetSize(qfinputfields[i], &size); CeedChk(ierr);
+      ierr = CeedVectorGetArray(impl->qvecsin[i], CEED_MEM_HOST, &tmp);
+      CeedChk(ierr);
+      ierr = CeedRealloc(numactivein + size, &activein); CeedChk(ierr);
+      for (CeedInt field=0; field<size; field++) {
+        activein[numactivein+field] = &tmp[field*Q];
+      }
+      numactivein += size;
+      ierr = CeedVectorRestoreArray(impl->qvecsin[i], &tmp); CeedChk(ierr);
```

Comment:
How do we know these arrays have been zeroed before we start using them below?

Also, I'm not wild about the action-at-a-distance of restoring the array. Probably better to do that down below, after using it.

Reply:
Added an explicit zero. I was just getting lucky by calling Assemble before OperatorApply in the tests.

We can't apply the QFunction without restoring access to these qvecs. I saw options as
1) Restore the array but keep a pointer (what I'm doing now)
2) Keep track of the vec and offset from the head of the array for each input field and use GetArray/RestoreArray on each loop

1 felt more straightforward and cleaner to me, but I'd be happy to switch to 2 if not bypassing the access system we set up is important.

Reply:
I guess it isn't a big deal at this point, just cruft in our model. Having some short scratch vectors into which you place the arrays might also work (if I'm context-switching back well enough that this isn't nonsense).